lib/mqtt.c

```diff
@@ -0,0 +1,473 @@
+  if(result)
+    goto fail;
+
+  /* fixme */
+  if(nread < MQTT_SUBACK_LEN) {
+    result = CURLE_WEIRD_SERVER_REPLY;
+    goto fail;
+  }
+
+  /* verify SUBACK */
+  if(readbuf[0] != 0x90 ||
+     readbuf[1] != 0x03 ||
+     readbuf[2] != ((mqtt->packetid >> 8) & 0xff) ||
+     readbuf[3] != (mqtt->packetid & 0xff) ||
+     readbuf[4] != 0x00)
+    result = CURLE_WEIRD_SERVER_REPLY;
+
+fail:
+  return result;
+}
+
+static int mqtt_encode_len(char *buf, size_t len)
+{
+  unsigned char encoded;
+  int i;
+
+  for(i = 0; len > 0; i++) {
+    encoded = len % 128;
+    len = len / 128;
+    if(len > 0)
+      encoded = encoded | 128;
+    buf[i] = encoded;
+  }
+
+  return i;
+}
+
+static CURLcode mqtt_publish(struct connectdata *conn)
+{
+  CURLcode result;
+  char *payload = conn->data->set.postfields;
+  size_t payloadlen = conn->data->set.postfieldsize;
+  char *topic;
+  size_t topiclen;
+  unsigned char *pkt;
+  size_t i = 0;
+
+  pkt = malloc(payloadlen + 10);
+  if(!pkt) {
+    result = CURLE_OUT_OF_MEMORY;
+    goto fail;
+  }
+
+  result = mqtt_get_topic(conn, &topic, &topiclen);
+  if(result)
+    goto fail;
+
+  /* assemble packet */
+  pkt[i++] = MQTT_MSG_PUBLISH;
+  i += mqtt_encode_len((char *)pkt + i, 2 + topiclen + payloadlen);
+  pkt[i++] = (topiclen >> 8) & 0xff;
+  pkt[i++] = (topiclen & 0xff);
+  memcpy(pkt + i, topic, topiclen);
+  i += topiclen;
+  memcpy(pkt + i, payload, payloadlen);
+  i += payloadlen;
+  result = mqtt_busy_write(conn, (char *)pkt, i);
+
+fail:
+  if(pkt)
+    Curl_safefree(pkt);
+  if(topic)
+    Curl_safefree(topic);
+
+  return result;
+}
+
+static size_t mqtt_decode_len(char *buf, size_t buflen, size_t *lenbytes)
+{
+  size_t len = 0;
+  size_t mult = 1;
+  size_t i;
+  unsigned char encoded = 128;
+
+  for(i = 0; (i < buflen) && (encoded & 128); i++) {
+    encoded = buf[i];
+    len += (encoded & 127) * mult;
+    mult *= 128;
+  }
+
+  *lenbytes = i;
+
+  return len;
+}
+
+static CURLcode mqtt_read_publish(struct connectdata *conn)
+{
+  CURLcode result;
+  curl_socket_t sockfd = conn->sock[FIRSTSOCKET];
+  ssize_t nread;
+  char *pkt = NULL;
+  size_t pktlen, lenbytes;
+  char *ptr;
+  size_t topiclen;
+  size_t payloadlen;
+  size_t packetidlen = 0;
+
+  /* allocate enough for a short message */
+  pkt = malloc(130);
+  if(!pkt) {
+    result = CURLE_OUT_OF_MEMORY;
+    goto fail;
+  }
+
+  /* packet header is 2-5 bytes long, read 5 to be sure we get all of it */
+  result = Curl_read(conn, sockfd, pkt, 5, &nread);
+  if(result) {
+    goto fail;
+  }
+
+  /* we are expecting a PUBLISH message */
+  if((pkt[0] & 0xf0) != 0x30) {
+    result = CURLE_WEIRD_SERVER_REPLY;
+    goto fail;
+  }
+
+  /* if QoS is set, message contains packet id */
+  if(pkt[0] & 6)
+    packetidlen = 2;
+
+  pktlen = mqtt_decode_len(pkt + 1, 4, &lenbytes);
+
+  /* reallocate a bigger buffer if necessary */
+  if(pktlen > 128) {
+    pkt = realloc(pkt, pktlen + 2); /* one extra for payload termination */
```

Comment:
`realloc()` without a temporary returnvalue will overwrite the original pointer on allocation error, better to use `Curl_saferealloc()` here instead.